src/mpid/ch4/src/ch4_vci.h

```diff
@@ -106,6 +106,25 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_map_contextid_rank_tag_to_vci(MPIR_Context_id
             % MPIDI_global.n_vcis);
 }
 
+/* Use round robin algorithm to assign vci_idx for transmit context */
+MPL_STATIC_INLINE_PREFIX int MPIDI_map_comm_to_vci_round_robin(MPIR_Comm * comm)
+{
+    MPID_THREAD_CS_ENTER(VCI, MPIR_THREAD_VCI_COMM_MUTEX(comm));
+    /* NOTE: Currently the round robin is done in the scope of a single communicator.
+     * Alternatively, this can be done at a global scope, across all the communicators.
+     */
+    if (unlikely(MPIDIG_COMM(comm, vci_idx_last_assigned) == MPIDI_VCI_INVALID)) {
+        MPIDIG_COMM(comm, vci_idx_last_assigned) = 0;   /* Assuming 1st idx is 0 */
+    } else {
+        int vci_idx_new = (MPIDIG_COMM(comm, vci_idx_last_assigned) + 1)
+            % MPIDI_global.n_vcis;
+        MPIDIG_COMM(comm, vci_idx_last_assigned) = vci_idx_new;
+    }
+
+    MPID_THREAD_CS_EXIT(VCI, MPIR_THREAD_VCI_COMM_MUTEX(comm));
+    return MPIDIG_COMM(comm, vci_idx_last_assigned);
+}
+
```

Comment:
I am not sure this is a good idea. How about split this commit to a separate PR so we can review it separately?

Reply:
Did you mean split just this one commit or the whole implementation to utilize the hint `allow_overtaking`? (this commit is a part of `allow_overtaking` based vci selection.)

Reply:
Let's split the entire `allow_overtaking` feature.

Reply:
Removed all three patches for `allow_overtaking`. Will create another PR for them.